examples/hal/quad/main.rs

```diff
@@ -4,613 +4,1357 @@
 )]
 
 extern crate env_logger;
-extern crate gfx_hal as hal;
 #[cfg(feature = "dx12")]
 extern crate gfx_backend_dx12 as back;
-#[cfg(feature = "vulkan")]
-extern crate gfx_backend_vulkan as back;
-#[cfg(feature = "metal")]
-extern crate gfx_backend_metal as back;
 #[cfg(feature = "gl")]
 extern crate gfx_backend_gl as back;
-
-extern crate winit;
+#[cfg(feature = "metal")]
+extern crate gfx_backend_metal as back;
+#[cfg(feature = "vulkan")]
+extern crate gfx_backend_vulkan as back;
+extern crate gfx_hal as hal;
 extern crate image;
+extern crate takeable_option;
```

Comment:
can we get away without using this?

Reply:
Lots gfx structs need to be moved into a destroy function like `destroy_descriptor_pool` when being destroyed. We can't move things out of structs (even in the Drop function) without putting something back in.

Now, if we had some empty default objects we could do std::mem::replace(old, default). Alternatively, we could use Option and plaster the code with unwraps (which I nor @msiglreith like) or we can use my thin wrapper around Option which has the Deref and DerefMut traits (which do the unwrapping for us). 

Of course, there might be other ways, but I can't think of them.

Reply:
Isn't this only a problem when you try to wrap things inside a struct that lives longer? The old example code just did
```rust
fn main() {
   let descriptor_pool = device.create_descriptor_pool(...);
   // do stuff
   device.destroy_descriptor_pool(descriptor_pool);
}
```

Reply:
Yes, only when they are in a structure. I want to avoid having one big function doing everything, and if we don't wrap related fields in structures then we might have to pass half a dozen to a dozen fields to everything.

Reply:
> I want to avoid having one big function doing everything

This is fine for an example, especially if it simplifies destruction.